tests/present_and_fetch.rs

```diff
@@ -0,0 +1,79 @@
+use softbuffer::{Context, Surface};
+
+use std::num::NonZeroU32;
+use std::panic::{catch_unwind, AssertUnwindSafe};
+
+use winit::event::Event;
+use winit::event_loop::{EventLoop, EventLoopWindowTarget};
+
+struct WinitBasedTest {
+    name: &'static str,
+    test_fn: fn(&EventLoopWindowTarget<()>),
+}
+
+fn all_red(elwt: &EventLoopWindowTarget<()>) {
+    let window = winit::window::WindowBuilder::new()
+        .with_title("all_red")
+        .with_visible(false)
+        .build(elwt)
+        .unwrap();
+
+    let context = unsafe { Context::new(elwt) }.unwrap();
+    let mut surface = unsafe { Surface::new(&context, &window) }.unwrap();
+    let size = window.inner_size();
+
+    // Set the size of the surface to the size of the window.
+    surface
+        .resize(
+            NonZeroU32::new(size.width).unwrap(),
+            NonZeroU32::new(size.height).unwrap(),
+        )
+        .unwrap();
+
+    // Set all pixels to red.
+    let mut buffer = surface.buffer_mut().unwrap();
+    buffer.fill(0xFF0000FF);
```

Comment:
That looks like blue to me.
https://github.com/rust-windowing/softbuffer/blob/fcda747ddfb7062606787a369540db5badf03503/src/lib.rs#L340
```suggestion
    buffer.fill(0x00FF0000);
```

Reply:
`0x00FF0000` won't be the same color on big-endian systems, right? If we care about those.

Something like `u32::from_ne_bytes` could be used for consistent behavior.

Reply:
I don't think so? I went down a rabbit hole for a bit there and confirmed two things:
1. Literals are always written in big-endian in Rust. I couldn't find any documentation on that, so I tried it out: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c1a0820eba008226bae7771c1f510fe8.
2. Bitshift operators always act in big-endian (except on some old PowerPC architectures I read somewhere?).

So I think what we are doing in Rust right now, seems to be pretty much cross-endianness compatible.
What I don't know is if the APIs expect us to turn things around on big-endian targets.
I also assume nobody here actually has a big-endian system lying around to try it out?

That said, on Web, my purview, this isn't a concern.

Reply:
I think the use of bitshifts in the examples would also be wrong on big endian. We represent color with `u32`, but it's really interpreted more like `[u8; 4]` but with 4 byte alignment. Which isn't dependent on endianess.

I don't have a big endian system to test on, though I fixed a similar issue in tiny-skia that someone reported: https://github.com/RazrFalcon/tiny-skia/pull/69. That one is easy to test with virtualization, but this is a bit more complicated since one may want a full OS with a display server. Maybe X11 backend over TCP could work...

Reply:
> Maybe X11 backend over TCP could work...

Sounds like fun, please share the result when you get them :rofl:.

Reply:
> 2\. Bitshift operators always act in big-endian

> I think the use of bitshifts in the examples would also be wrong on big endian.

I think these two contradict each other, if bitshift operators are always acting in big-endian, then it shouldn't matter what platform you are on. Unless, looking at your PR, you were using PowerPC to test: 
> (except on some old PowerPC architectures I read somewhere?)

Or maybe my assumptions are just wrong, I can't really test them.

Reply:
I don't think "bitshift operators are always acting in big-endian" really makes sense? That would mean `1 << 3` for instance isn't `8`, because when compared in native endianess it is interpreted as something other than `8`?

It looks like you're right that there's something weird about how bitshifts and endianness work on powerpc, at the assembly level, but at the Rust or C level they should behave the same way across platforms in how they operate on integer values, while the actual order of the bytes in memory may differ.

Reply:
> I don't think "bitshift operators are always acting in big-endian" really makes sense? That would mean `1 << 3` for instance isn't `8`, because when compared in native endianess it is interpreted as something other than `8`?

No, what I'm saying is that `1 << 3` is always `8`, no matter the endianness of the platform. That is why I think any code we are using here should be right.

... unless of course the API isn't really seeing a `u32`, but a `[u8; 4]`, in which case I would think that the same number would have a different meanings depending on the platform endianness.

> We represent color with `u32`, but it's really interpreted more like `[u8; 4]` but with 4 byte alignment. Which isn't dependent on endianess.

... contrary to what you said earlier.

(lots of emphases on "I think", again: can't test any of my assumptions here)

Reply:
I mean that the pixel format is basically a `[u8; 4]`, or "big-endian" BGRX, (B byte, G byte, R byte, then X byte) regardless of of platform endianess. So the *byte order* of the pixels doesn't depend on endianness, but consequently the integer value this corresponds to will depend on endianess.

I guess to test this I'll need to cross compile at least libxcb for big endinan powerPC or mips. I haven't had luck getting either to boot in a VM, though if I can cross compile a binary qemu user emulation should work...

But this isn't that important if it's hard to even find something big endian to emulate. It would just be nice to do the correct thing here.

Reply:
> I mean that the pixel format is basically a `[u8; 4]`, or "big-endian" BGRX, (B byte, G byte, R byte, then X byte) regardless of of platform endianess. So the _byte order_ of the pixels doesn't depend on endianness, but consequently the integer value this corresponds to will depend on endianess.

This sounds very correct to me.

So yeah, agreed, this would need correction.

---

On that note, why did we decide to use `u32` instead of `[u8; 4]`?

Reply:
It can be faster to work with 32-bit values than individual bytes. And things like tiny-skia work on `&[u32]`. And you can safely cast `&[u32]` to `&[u8]` but not the other way around due to alignment.

Reply:
I'm not sure what to think of this, there isn't actually a **safe** method to cast `&[u32]` to `&[u8]` I know of. I would argue that `u32` is definitely a worse user interface and the improved speed is speculation (I actually agree with that speculation).

Just an idea: what do you think of `&[[u8; 4]]`?

Reply:
`[u32]` has alignment `4` (on typical architectures), while `[u8; 4]` has alignment `1`. So there's not currently a standard library method to cast `&[u32]` to `&[u8]`, but it's always sound while the reverse is not true. Likewise it is potentially unsound to cast `&[[u8; 4]]` to `&[u32]`.

`bytemuck` is popular for this sort of thing.

Reply:
> while `[u8; 4]` has alignment `1`.

Ah, apologies, was thinking that this does have an alignment of 4 actually, for some reason (I think I played around with too much shaders lately ...).

---

Alright, w/e, I think I'm still in favor of `&[u8]` until we can show a reasonable performance degradation, but this is an issue for another time.

In the meantime, your original problem still needs to be addressed. I think we should also document this somewhere.
https://github.com/rust-windowing/softbuffer/blob/fcda747ddfb7062606787a369540db5badf03503/src/lib.rs#L340
This could be a bit confusing, because it's showing big-endian order, which most platforms aren't, but it's also representative of how literals are written in Rust.

Reply:
I say we handle it afterwards: https://github.com/rust-windowing/softbuffer/issues/109.